react-sdk/src/providers/tambo-context-attachment-provider.tsx

```diff
@@ -0,0 +1,281 @@
+ *   <App />
+ * </TamboProvider>
+ * ```
+ */
+export function ContextAttachmentProvider({
+  children,
+  getContextHelperData,
+}: ContextAttachmentProviderProps) {
+  const [attachments, setAttachments] = useState<ContextAttachment[]>([]);
+  const [customSuggestions, setCustomSuggestions] = useState<
+    Suggestion[] | null
+  >(null);
+  const { addContextHelper, removeContextHelper } = useTamboContextHelpers();
+
+  // Track which context helpers have been registered to avoid duplicates
+  const registeredIdsRef = useRef<Set<string>>(new Set());
+  const prevGetContextHelperDataRef = useRef<typeof getContextHelperData>();
+
+  // Sync context helpers with attachments using useEffect
+  useEffect(() => {
+    const currentIds = new Set(attachments.map((a) => a.id));
+    const registeredIds = registeredIdsRef.current;
+
+    // Remove context helpers that are no longer in attachments
+    registeredIds.forEach((id) => {
+      if (!currentIds.has(id)) {
+        removeContextHelper(id);
+        registeredIds.delete(id);
+      }
+    });
+
+    const getDataChanged =
+      prevGetContextHelperDataRef.current !== getContextHelperData;
+
+    // Add or replace context helpers for attachments
+    attachments.forEach((context) => {
+      if (getDataChanged || !registeredIds.has(context.id)) {
+        addContextHelper(context.id, async (): Promise<ContextHelperData> => {
+          if (getContextHelperData) {
+            return await getContextHelperData(context);
+          }
+          return {
+            selectedComponent: {
+              name: context.name,
+              instruction:
+                "This is a Tambo interactable component that is currently selected and visible on the dashboard. You can read its current props and state, and update it by modifying its props. If multiple components are attached, you can interact with and modify any of them. Use the auto-registered interactable component tools (like get_interactable_component_by_id and update_interactable_component_<id>) to view and update the component's state.",
+              ...(context.metadata ?? {}),
+            },
+          };
+        });
+        registeredIds.add(context.id);
+      }
+    });
+
+    prevGetContextHelperDataRef.current = getContextHelperData;
+  }, [
+    attachments,
+    addContextHelper,
+    removeContextHelper,
+    getContextHelperData,
+  ]);
+
+  // Cleanup: remove all context helpers on unmount
+  useEffect(() => {
+    const registeredIds = registeredIdsRef.current;
+    return () => {
+      registeredIds.forEach((id) => {
+        removeContextHelper(id);
+      });
+      registeredIds.clear();
+    };
+  }, [removeContextHelper]);
+
+  const addContextAttachment = useCallback(
+    (context: Omit<ContextAttachment, "id">) => {
+      setAttachments((prev) => {
+        if (prev.some((c) => c.name === context.name)) return prev;
+
+        const newId =
+          typeof crypto !== "undefined" && "randomUUID" in crypto
```

Comment:
I think we need to expect crypto to be there - and throwing an error is actually the right move if it isn't.  I did learn that the reason it sometimes isn't present is because you're using an ip address instead of localhost